### ISSUE TITLE
Incremental adding in indie-api…?

### ISSUE THREAD

**Issue by hedefalk** [steelbrain/linter#1087]
I don't seem to understand how I can use the indie-api to support our use case. We have errors streaming in from a backend server and I want to add them as they come.

Indie api only supports `setMessages(array)` so I figured I had to do my own stateful array to push to. However, that made some caching in linter go bananas and I got the explanation that it always needs to be new objects:

https://github.com/atom-community/linter/issues/1075#issuecomment-181911983

The quick workaround was to just do `setMessages(JSON.parse(JSON.stringify(messages)))``

If I do this on every incoming new message I'll have O(n^2) calls to JSON parse (triangle number).

On 1000 messages, this is obviously pretty bad:

https://github.com/ensime/ensime-atom/issues/187

Even if we can to better than using the JSON methods I see no feasible workaround if we need to throw in new objects all the time.

I see three options:

* An `addMessages(messages)` to add incrementally to linter api
* Remove that thing that makes needing new objects so we can reuse the ones we have
* That we on the client side do scheduling and only make calls to `setMessages()` when we think server has stopped streaming us errors.

Obviously I'd prefer not to do the last one :)

Any other suggestions or tip on ways to do this are highly appreciated!

**Comment by steelbrain** [comment]
This has already been fixed in master, gonna release soon

**Comment by hedefalk** [comment]
@steelbrain Aaawesome! Please let me know when :)

**Comment by steelbrain** [comment]
By the pace of progress, about 15 days give or take

**Comment by hedefalk** [comment]
@steelbrain Oh, I think I need to know how it is fixed so I can make the client fix adherently. Currently, no fix in linter will make the performance problem in ensime-atom go away since it's our code that is hogging CPU.

**Comment by hedefalk** [comment]
Ok, so I can't find any addMessages in master so I guess the uniqueness constraint has been lifted…?

**Comment by steelbrain** [comment]
Yeah the uniqueness has been fixed, you can reuse the same objects

**Comment by steelbrain** [comment]
We're also gonna be rolling out a different kind-of API in the near future that will look like
```js
myLinter.setMessage({type: 'Error', ...}, 'some-unique-key')
myLinter.invalidateMessage('some-unique-key')
```

**Comment by hedefalk** [comment]
Cool! Looking forward!

**Comment by hedefalk** [comment]
I guess the non-uniqueness contstraint was lifted in 1.11.4 so I'll fixup with

-        indieLinter.setMessages(JSON.parse(JSON.stringify(lints)))
+        indieLinter.setMessages(lints)

and close this one?

**Comment by steelbrain** [comment]
Well actually, it was just a patch release with fixes for latest version of electron

**Comment by hedefalk** [comment]
@steelbrain Ok, so the diff above will make linter usage go super-bad again?

**Comment by steelbrain** [comment]
Yep, sorry about that, forgot to include this fix in that path release

**Comment by hedefalk** [comment]
Ok, cool, I'll revert!

**Comment by steelbrain** [comment]
Fixed in v1.11.5
